### PR TITLE
Re-enable RMC to allow users to verify non-public main.

### LINKED ISSUE #655: Re-enable support to binary crates verification
We should be able to verify bin crates when the verification target is the main function. I.e., user should be able to verify the following program:

```rust
// my_main.rs
fn main() {
    // code goes here.
}
```

using the following command line invocation:

```
rmc my_main.rs
```

with RMC version:

I expected to see this happen: RMC verifies the main function.

Instead, this happened: RMC cannot find main function because it is not marked as public.

### PR DESCRIPTION
**Pull request by celinval**
Users don't usually set their main function to be public and we should enable them to verify their main function. For now, users will have to manually set the create type when running RMC directly. For cargo rmc, this should be automatically set up by cargo.

### Resolved issues:

Resolves #655 

### Call-outs:

We should figure out what is the default behavior of RMC and we might want to add a mechanism to figure out which one to use based on the user configuration. For example, maybe we can check what is the target function for the verification.

### Testing:

* How is this change tested?

* Is this a refactor change?

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/tools/dashboard/src/books.rs [hunk 1]
```diff
@@ -296,17 +291,15 @@ fn extract(par_from: &Path, par_to: &Path, config_paths: &mut HashSet<PathBuf>)
...
-                None,
-            )
-            .0,
-        );
+        example.code = rustdoc::doctest::make_test(
+            &example.code,
+            None,
+            false,
+            &Default::default(),
+            example.config.edition.unwrap_or(Edition::Edition2018),
```
Comment (adpaco-aws) [inline src/tools/dashboard/src/books.rs:299]: The default edition was set to 2021 in #652, so this should be:
```suggestion
            example.config.edition.unwrap(),
```

File: src/test/expected/one-assert/test.rs [hunk 2]
```diff
@@ -1,7 +1,9 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0 OR MIT
-
-pub fn main() {
+// rmc-flags: --function check_assert
+// compile-flags: --crate-type lib
```
Comment (adpaco-aws) [inline src/test/expected/one-assert/test.rs:4]: Why is this needed?

File: src/test/rmc/ArithOperators/main.rs [hunk 3]
```diff
@@ -1,6 +1,6 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0 OR MIT
-pub fn main() {
+fn main() {
```
Comment (adpaco-aws) [inline src/test/rmc/ArithOperators/main.rs:3]: Isn't this change needed for each test as in #523?

File: src/tools/compiletest/src/runtest.rs [hunk 4]
```diff
@@ -2414,16 +2410,7 @@ impl<'test> TestCx<'test> {
...
-        // Arguments to choose specific abstraction are currently provided as
-        // as rmc-flags in the test file
-        rmc.args(&self.props.rmc_flags)
-            .arg("--input")
-            .arg(&self.testpaths.file)
-            .arg("--cbmc-args")
-            .args(&self.props.cbmc_flags);
-        self.add_rmc_dir_to_path(&mut rmc);
-        let proc_res = self.compose_and_run_compiler(rmc, None);
+        let proc_res = self.run_rmc();
```
Comment (adpaco-aws) [inline src/tools/compiletest/src/runtest.rs:2413]: Thanks for refactoring this!

**Review by adpaco-aws — CHANGES_REQUESTED** [review]
Does not look ready to me, see comments below. Also, how it comes we are able to codegen binary crates now? The changes in RMC scripts remove all lib-related flags, I do not see any major change anywhere in this PR.

File: src/test/expected/one-assert/test.rs (hunk 2 above)
Comment (celinval) [reply src/test/expected/one-assert/test.rs:4]: This is needed because binary crates have more checks than library crates. The purpose of this test is to ensure that there is only one assertion in the result.

File: src/test/rmc/ArithOperators/main.rs (hunk 3 above)
Comment (celinval) [reply src/test/rmc/ArithOperators/main.rs:3]: No. It's not needed as in #523. With this change RMC will accept both public and non public main.

**Comment by celinval** [comment]
> Does not look ready to me, see comments below. Also, how it comes we are able to codegen binary crates now? The changes in RMC scripts remove all lib-related flags, I do not see any major change anywhere in this PR.

Previous fixes is what enabled us to be able to codegen binary crates. This PR is just to remove the hack that we were enforcing the crate type to be a lib.

File: src/test/rmc/ArithOperators/main.rs (hunk 3 above)
Comment (adpaco-aws) [reply src/test/rmc/ArithOperators/main.rs:3]: Yes, I understand it is not strictly need but nice to have. Anyway, let me take care of that in #659

**Review by adpaco-aws — APPROVED** [review]

File: src/test/rmc/ArithOperators/main.rs (hunk 3 above)
Comment (celinval) [reply src/test/rmc/ArithOperators/main.rs:3]: Thanks

File: src/test/expected/one-assert/test.rs (hunk 2 above)
Comment (zhassan-aws) [reply src/test/expected/one-assert/test.rs:4]: I'm curious what the extra checks for binary crates are.

**Review by zhassan-aws — APPROVED** [review]

File: src/tools/compiletest/src/runtest.rs [hunk 5]
```diff
@@ -2273,26 +2273,7 @@ impl<'test> TestCx<'test> {
...
-        }
-        // Pass the test path along with RMC and CBMC flags parsed from comments at the top of the test file.
-        rmc.args(&self.props.rmc_flags)
-            .arg("--input")
-            .arg(&self.testpaths.file)
-            .arg("--cbmc-args")
-            .args(&self.props.cbmc_flags);
-        self.add_rmc_dir_to_path(&mut rmc);
-        let proc_res = self.compose_and_run_compiler(rmc, None);
+        let proc_res = self.run_rmc();
```
Comment (danielsn) [inline src/tools/compiletest/src/runtest.rs:2276]: 💯
